apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java

```diff
@@ -165,20 +173,35 @@ public List<NamespaceBO> findNamespaceBOs(String appId, Env env, String clusterN
       throw new BadRequestException("namespaces not exist");
     }
 
-    List<NamespaceBO> namespaceBOs = new LinkedList<>();
+    List<NamespaceBO> namespaceBOs = new CopyOnWriteArrayList<>();
+    List<String> exceptionNamespaces = new CopyOnWriteArrayList<>();
+    CountDownLatch latch = new CountDownLatch(namespaces.size());
     for (NamespaceDTO namespace : namespaces) {
+      executorService.submit(() -> {
+        NamespaceBO namespaceBO;
+        try {
+          namespaceBO = transformNamespace2BO(env, namespace);
+          namespaceBOs.add(namespaceBO);
+        } catch (Exception e) {
+          LOGGER.error("parse namespace error. app id:{}, env:{}, clusterName:{}, namespace:{}",
+              appId, env, clusterName, namespace.getNamespaceName(), e);
+          exceptionNamespaces.add(namespace.getNamespaceName());
+        } finally {
+          latch.countDown();
+        }
+      });
 
-      NamespaceBO namespaceBO;
-      try {
-        namespaceBO = transformNamespace2BO(env, namespace);
-        namespaceBOs.add(namespaceBO);
-      } catch (Exception e) {
-        LOGGER.error("parse namespace error. app id:{}, env:{}, clusterName:{}, namespace:{}",
-            appId, env, clusterName, namespace.getNamespaceName(), e);
-        throw e;
-      }
+    }
+    try {
+      latch.await();
+    } catch (InterruptedException e) {
+      //ignore
     }
 
+    if(namespaceBOs.size() != namespaces.size()){
+       throw new RuntimeException(String
+           .format("Parse namespaces error, expected: %s, but actual: %s, cannot get those namespaces: %s", namespaces.size(), namespaceBOs.size(), exceptionNamespaces));
```

Comment:
The original logic will throw the actual exception so that the user knows what's wrong. Is it better to preserve that logic? For example, we could use future.get to retrieve the exception occurred.

```java
    List<NamespaceBO> namespaceBOs = Collections.synchronizedList(new LinkedList<>());
    List<Future<?>> futures = new LinkedList<>();
    for (NamespaceDTO namespace : namespaces) {
      futures.add(executorService.submit(() -> {
        NamespaceBO namespaceBO;
        try {
          namespaceBO = transformNamespace2BO(env, namespace);
          namespaceBOs.add(namespaceBO);
        } catch (Exception e) {
          LOGGER.error("parse namespace error. app id:{}, env:{}, clusterName:{}, namespace:{}",
              appId, env, clusterName, namespace.getNamespaceName(), e);
          throw e;
        }
      }));
    }

    futures.forEach(future -> {
      try {
        future.get();
      } catch (ExecutionException | InterruptedException e) {
        Throwable exception = e instanceof ExecutionException ? e.getCause() : e;
        if (exception instanceof RuntimeException) {
          throw (RuntimeException) exception;
        } else {
          throw new RuntimeException(exception);
        }
      }
    });
```

Reply:
```java
LOGGER.error("parse namespace error. app id:{}, env:{}, clusterName:{}, namespace:{}",
              appId, env, clusterName, namespace.getNamespaceName(), e);
```
The original exception stack information will be recorded here

Reply:
I understand they are recorded in the server logs. What I propose is to throw the detailed exception message to the UI page so that the operating users could understand what's wrong, which is the original behavior.
The current error message `Parse namespaces error, expected: %s, but actual: %s, cannot get those namespaces: %s` could show the operating users some error details, but I have a concern that it might not be very clear.

Reply:
I probably see what you mean, and the main points I took into account when making the changes here were.
- I am not very clear about the internal logic will report what possible exceptions, but when I simulated an exception, as a direct user it is difficult to do anything based on this exception, may be my simulation and the real difference
- the original will only encounter a Namespace problem to report an exception, now all the problems will be reported at once, will it be better?

Reply:
> I am not very clear about the internal logic will report what possible exceptions, but when I simulated an exception, as a direct user it is difficult to do anything based on this exception, may be my simulation and the real difference

The exceptions would be either timeout exceptions occurred in the portal side, or database access exceptions occurred in the admin service side. I suppose there is not much to do for a direct user, except that the user could report that to the admin users.
So my original idea was that we keep the logic the same as before - throw the actual exception and display it in the UI side, but I think the current logic is also ok.

> the original will only encounter a Namespace problem to report an exception, now all the problems will be reported at once, will it be better?

That I suppose is the same as the user can't really handle the errors, no matter it contains one error or multiple errors :-)